### ISSUE TITLE
default_factory values leaking into compiled schema

### ISSUE THREAD

**Issue by michal-rogala** [marcosschroh/dataclasses-avroschema#400]
Non-optional field with default_factory leaks default value into compiled schema:

```
from uuid import uuid4, UUID

from dataclasses_avroschema.avrodantic import AvroBaseModel
from pydantic import Field

class TestSchema(AvroBaseModel):
     uuid: UUID = Field(default_factory=uuid4)

print(TestSchema.avro_schema())
```

results in schema:

```
{
  "type": "record",
  "name": "TestSchema",
  "fields": [
    {
      "name": "uuid",
      "type": {
        "type": "string",
        "logicalType": "uuid"
      },
      "default": "8671a831-9cf5-4875-b0cf-5ed7aad81a95"
    }
  ]
}
```

As default value is dynamically generated - it makes no sense to have it statically compiled into schema. Also - does having "default" value baked in, make the field optional from application point of view?
cc: @ddevlin

**Comment by marcosschroh** [comment]
@michal-rogala that is the expected behavior. You should not do that. If you want a default value which is a default uuid just return the `string` or you can have a default factory that returns always the same result

**Comment by michal-rogala** [comment]
What we want to achieve is to have application automatically fill non-optional values - like UUID or timestamps (which cannot be static by design) when object is being generated. Those values should not propagate to schema "default" field because it does not make sense in that case (like timestamp). One way to work around this to fill those values in custom constructor but it does not scale with objects with thousands nested fields.

Do you have an idea how to achieve that? Maybe we could implement annotation to field to skip adding those values to schema.

**Comment by marcosschroh** [comment]
@michal-rogala Ah now I understand what you need. There is not a proper way to do it, but:

Do you register the schema every time that you encounter a new class with  a different  default  value? I am asking because even though the default value changes every time still should be possible to `serialize/deserialize` `payloads/events` with schemas that are the same except the default value. What do you do with the schemas?

**Comment by marcosschroh** [comment]
You're welcome

>  I was wondering if this behaviour is normal - passing factory as default value implies value is dynamic (uuid/timestamp, etc) and putting it into default value in schema seemed odd.

I agree that is odd but if you do not register a schema every time then it is fine. Even if the field has a default value in the schema it won't be a problem because you also have it in the `default_factory` and its output always will be used to serialize the event. 

If you register a schema, then having a default value on fields like updated_on (`timestamp`) it's weird but possible, everything dependens on  the use case. I will try to write some documentation about it.

Let's close this issue for now. If you or someone else experience another issue or believes that this is not the right way to go let's re open it.

**Comment by marcosschroh** [comment]
We can add an extra option to the `Meta` called `List[str]: exclude_default`. This option will contain the name of the fields whose default values should be omitted(`default` or `default_factory`)

**Comment by michal-rogala** [comment]
yes, this would would work - or an option to define it at field level - if you have nested dataclasses it would be difficult to manage Meta class.

**Comment by joaoe** [comment]
> We can add an extra option to the Meta called List[str]: exclude_default

That sounds good but it would make much more sense to configure that in the Field, whether the default_factory value can be exported.

**Comment by tribunsky-kir** [comment]
Hello! I've encountered the same issue and found this discussion helpful. 😉

>I agree that is odd but if you do not register a schema every time then it is fine.

That's true, that everything goes smoothly if schema is published only once.

However, in my own (albeit unrepresentative) experience with living in Kubernetes or its analogs, this behavior will just continuously bump the schema version to an excessive level. This is because pods are supposed to be restarted frequently, and each restart will publish a **new** default value.

Moreover, the same behavior occurs for any restarts of the process with the producer (e.g., local runs against a development cluster). In the case of Kafka, the producer has to push its schema to the schema registry after initialization to get the schema ID. As far as I know, this behavior is default for Java producers and the confluent-kafka-python producer.

So, I agree with @michal-rogala.

@marcosschroh, what is the specific use case for utilizing the value from the factory? In my past projects, I used `default_factory` specifically to dynamically generate a value 'out of thin air' in my code while keeping the schema unchanged.

Would it be better to retain the option to initialize the default value for the schema from the factory, but turn it off by default?

**Comment by marcosschroh** [comment]
Hi all, 

Sorry for the delay. The reason to have `default_factory` due to the need to have `default` values for `map, array and records`. 

Example:

Have a schema with `favourites_numbers` field that has a default value `[7, 13]` (it needs to be generated using the `default_factory` because it is a mutable value):

```python
@dataclasses.dataclass
class UserAdvance(AvroModel):
    ...
    favourites_numbers: typing.List[int] = dataclasses.field(default_factory=lambda: [7, 13])
```

resulting in:

```json
{
  "type": "record",
  "name": "UserAdvance",
  "fields": [
    ...
    {
      "name": "favourites_numbers",
      "type": {
        "type": "array",
        "items": "long",
        "name": "favourites_number"
      },
      "default": [7, 13]
    }
  ],
}
``` 

Then the question is: Should `default_factory` be applied only for `map, array and records` and any other mutable value?
I think yes, then it will solve this issue. 

If end users want to have a default value for `map, array and records` they have to use the `default_factory`  as `default` can not be used. It is likely that  the default will always be the same. For the other cases, `default` must be used.

Because it is not possible to use `default` and `default_factory` at the same time, we can use it as a distinction to include or not a default in the schema (except for mutable values). 

Then:

- For types like `date`, `time`, `datetime` and `uuid` that contain a default value which was generated using `default_factory` (they will change every time), they won't be included in the resulting schema but new instance will have it.
- 
Example:

```python
@dataclasses.dataclass
class UUIDLogicalTypes(AvroModel):
    event_uuid: uuid.UUID = dataclasses.field(default_factory=uuid.uuid4)


In [9]: UUIDLogicalTypes()
Out[9]: UUIDLogicalTypes(event_uuid=UUID('c0c48c38-6cf1-4fb8-ab13-dd3e9563dfd9'))
```

The resulting schema must be:

```json
"type": "record",
  "name": "UUIDLogicalTypes",
  "fields": [
    {
      "name": "uuid_1",
      "type": {"type": "string", "logicalType": "uuid"}
    }
  ]
}
```

- If a default value is needed in the schema, then `default` must be used. This will result that the schema won't change and the instances will have always the same default (this might not have any sense at all, but it is possible)

```python
@dataclasses.dataclass
class UUIDLogicalTypes(AvroModel):
    event_uuid: uuid.UUID = uuid.UUID('c0c48c38-6cf1-4fb8-ab13-dd3e9563dfd9')
```

What do you think?

**Comment by tribunsky-kir** [comment]
Thank you for the detailed answer!

Now I see what the case is: `dataclasses` disallow mutable default items, unlike Pydantic (my bad, I don't use them often, when I need a lightweight structure I still use named tuples).

Then, applying `default_factory` only for maps, arrays, and records seems to be more than a reasonable default (and a necessary evil). 

And I am strongly agree on types like `date`, `time`, `datetime` and `uuid`: as their default factories are not 'pure functions' and changes output between calls, they shouldn't be used in schema to define the default value and the default value must be passed implicitly.

My only concern then remains whether the same behavior should be applied to `AvroBaseModel`, where we can just define mutable defaults. Ideally, it would be the same as for `dataclasses` - it is easier to understand from the user's perspective. On the other hand, technically, it is not necessary to impose such restrictions: when the default value is mutable, it can still be defined via `Field`, and the schema generator should not consider it as a special case.

```python
from typing import List

from dataclasses_avroschema import AvroBaseModel
from pydantic import Field


# This just works and should be passed to schema
class PydanticUserAdvance(AvroBaseModel):
    ...
    favourites_numbers: List[int] = Field(default=[7, 13])

...

# At first glance, it should work as in dataclasses, but why schema generator should consider special cases 
# to skip/pass default-factory's value to the schema?
class PydanticUserAdvance(AvroBaseModel):
    ...
    favourites_numbers: List[int] = Field(default_factory=lambda: [7, 13])
```

What are your thoughts on this?

**Comment by marcosschroh** [comment]
I think it won't be a problem 

**Dataclasses**:

1. If end users want to define a default in the schema, then the `default` must be used except for `maps, array and records`. 
2. If `default_factory` is used for a type that is not a `maps, array or records` then it will be ignored in the final schema

**Pydantic**:

1. If end users want to define a default in the schema, then the `default` must be used. This also apply for `maps, array and records` unlike `dataclasses point 1` because with `pydantic` it is possible have mutable values as default.
2. If `default_factory` is used for a type that is not a ` maps, array or records` then it will be ignored in the final schema like in `dataclasses`.

This leads us to:

Usually, for `date, time, datetime or uuid` types  end user want to use the `default_factory` so instances will have auto generated values, then the value will be ignored in the final schema (restriction 2). If end users want a default value in the schema for these types, then `default` must be used (restriction 1), meaning that every instance will have the same value and it will match the python class.

For  `maps, array or records` is not a problem because the work only with restriction 2. If the default factory is dynamic then the schema will change as well, typically it is not the case.

**Comment by marcosschroh** [comment]
Hi all,

I tried to follow the approach that I describe `above` but I realized that it does not scale,  we have users that generated the schema once and then they use it in a different programing language, meaning that we can add custom behavior/rules to specific types. Then, the only solution is that the end user must specifies when a `default` value should be `excluded`. 

I followed the [marshmallow_dataclass](https://github.com/lovasoa/marshmallow_dataclass) approach, which is what @joaoe suggested. It will be possible to use the config `exclude_default` in the `metadata`. Here is the [PR](https://github.com/marcosschroh/dataclasses-avroschema/pull/513), you can also take a look the [example in the documentation](https://marcosschroh.github.io/dataclasses-avroschema/pr-preview/pr-513/fields_specification/#exclude-default-value-from-schema)

**Comment by zhavir** [comment]
hello, is it possible to use the same feature with pydantic? I've tried but without success

**Comment by pillmuncher** [comment]
I just ran into that problem when I wanted to register a schema automatically with Confluent Schema Registry if it wasn't already registered. It took me a while to understand that it was the `default_factory=uuid.uuid4`.

I don't really agree with the argument that was given, that the schema should include the value generated by `default_factory`, so that mutable lists cannot become default values because they'd then be shared by all objects of that class. That's just how Python works. People should know that and not do such a thing. One of Python's maxims is "we're all consenting adults" and if people want to shoot themselves in the foot, we should let them. At least then they will be able to find an answer on StackOverflow to why their lists are shared among their objects, as opposed to what I had to do, where finding an answer as to what was going on was kinda hard.

So, no, I don't think it's a wise decision to bolt on some extra logic that runs counter to what Python *is*.

**Comment by marcosschroh** [comment]
The `default` value coming from the `default_factory` will be always included unless you specify the contrary: https://marcosschroh.github.io/dataclasses-avroschema/fields_specification/#exclude-default-value-from-schema

**Comment by pillmuncher** [comment]
Yes, I understand that. I also had found the solution already. Yet, I still think you have it backwards. According to this:

https://docs.python.org/3/library/dataclasses.html#mutable-default-values

you're changing documented and expected behavior. Again, I doubt that's a wise decision.

**Comment by marcosschroh** [comment]
We are not changing any behavior, on the contrary. If I have the following `dataclass`, what do you expect? 

```python
@dataclass
class D:
    x: list = field(default_factory=list)
```

As the documentation says `assert D().x is not D().x` (which is how it should be). 

Now, what can you do with this library? You can create `avro schemas`, to do it we need to inspect all the values using `dataclasses.fields` which it gives us all the `FIelds` with their properties (including `default_factory` which you do not know until you call it). If you call `D.avro_schema()` what do you think is going to happen? of course `assert  D.avro_schema() x is not  D.avro_schema()` because you are specifying a value that is changing every time, which it matches the `behavior` of instances creation.

**Comment by pillmuncher** [comment]
I still believe you got it wrong. Look:
```python
>>> from dataclasses import dataclass, field
>>> from datetime import datetime
>>> @dataclass
... class C:
...     d: datetime = field(default_factory=datetime.now)
... 
>>> c1 = C()
>>> c2 = C()
>>> c1 == c2
False
>>> type(c1) == type(c2)
True
>>> c1.d is not c2.d
True
>>> assert c1.d is not c2.d
```

See, the type does not change just because the `default_factory` returns a different value each time it's called. BTW, that's the whole point of having a `default_factory` in the first place: To return a different value each time it is called. Otherwise we could just write down the value literally and forget about calling a function. The Python docs even spell it out:

> Using default factory functions is a way to create new instances of mutable types as default values for fields:

```python
@dataclass
class D:
    x: list = field(default_factory=list)

assert D().x is not D().x

```
"_create new instances of mutable types_" - that means that the field values are mutable, not that the return value of a `default_factory` is part of the type, or the schema.
